### ISSUE TITLE
Experiments Strategy

### ISSUE THREAD

**Issue by jorahn** [jorahn/rookworld-trl#8]
@claude Check experiment*.md and potentially relevant code.
Methodically analyze what was testes, the parameters used and the results achieved.

Create a plan for hypotheses to test and experiments to run to get really meaningful insights quickly with the least code changes.

**Comment by jorahn** [comment]
# Comprehensive Hyperparameter Sweep Results & Overfitting Discovery (2025-09-19)

## Summary of Today's Work

Completed systematic hyperparameter exploration with **72 total training runs** across two sweeps:
1. **48-run broad sweep** (50 steps each) - initial parameter space exploration
2. **24-run focused sweep** (100 steps each) - refined ranges based on "high performers"

## Key Finding: Overfitting Problem Discovered

### Critical Analysis Error Correction
Initially reported "phenomenal success" based on training metrics, but **actual evaluation performance worsened**:

| KPI | Original 48-Run | Focused 24-Run | Change |
|-----|----------------|----------------|--------|
| **Eval Accuracy Success Rate** | 16.7% (8/48) | **8.3%** (2/24) | **-8.4pp WORSE** |
| **Eval Reward Success Rate** | 37.5% (18/48) | **12.5%** (3/24) | **-25pp MUCH WORSE** |
| **Avg Eval Accuracy Change** | -1.0% | **-1.6%** | **Worse generalization** |
| **Avg Eval Reward Change** | -0.010 | **-0.019** | **Worse quality** |

### Root Cause: Training vs Evaluation Divergence
- **Training metrics**: Consistent improvement (+0.2025 across all focused runs)
- **Evaluation metrics**: Degradation (-1.6% accuracy, -0.019 reward)
- **Diagnosis**: Model optimizing for training distribution, losing generalization ability

## Actual Insights from 72 Runs

### Best Evaluation Performance Parameters
Based on runs that actually improved eval accuracy (+2%):

```bash
# Conservative parameters for eval performance
--learning_rate 1.2e-07      # Range: 1.0-2.4e-07 (conservative end)
--lr_schedule cosine         # Slightly better for eval vs advanced
--entropy_coef 0.001         # Low exploration, better generalization  
--batch_size 8               # Confirmed optimal (100% of improvers)
--gens 16                    # Moderate GRPO signal
--steps 100-150              # Sufficient for convergence
--eval_every 10              # Critical for overfitting detection
```

### Learning Rate Sensitivity (Corrected)
- **1e-7 to 2.4e-07**: Only range showing eval accuracy improvements
- **>1e-6**: Still harmful for both training AND eval metrics
- **Ultra-conservative**: May be needed for generalization vs training performance

## Strategic Implications

### Hyperparameter Optimization: Limited Returns
- **72 runs of evidence**: Parameter tweaking shows diminishing returns
- **Eval ceiling**: ~70% maximum achieved across all configurations
- **Overfitting risk**: Tighter parameter control can hurt generalization

### Next Priority: Fundamental Improvements
1. **Curriculum Learning**: Focus on Move 1 (only consistent improvement area)
2. **Model Capacity**: Test if 124M parameters are insufficient for 80% target
3. **Dataset Quality**: Analyze training/eval composition differences
4. **Generation Quality**: Ensure training improvements represent real chess knowledge

## Infrastructure Achievements

### Multi-GPU Sweep Framework
- **Parallel execution**: 2 GPUs × 4 concurrent = 8x speedup
- **Memory management**: Safe VRAM usage with gradient update accounting
- **Systematic analysis**: Proper KPI tracking with eval-focused metrics

### Methodological Value
- **Overfitting detection**: Systematic framework for identifying training/eval divergence
- **Evidence-based refinement**: Data-driven parameter optimization methodology
- **KPI clarity**: Importance of focusing on actual target metrics (eval accuracy/reward)

## Recommended Next Experiments

### Priority 1: Curriculum Learning (High Impact)
```bash
# Move 1 focused training with conservative parameters
uv run python manual_grpo_debug.py \
  --learning_rate 1.2e-07 \
  --lr_schedule cosine \
  --entropy_coef 0.001 \
  --steps 200 \
  --eval_every 20 \
  --task_type A
```

### Priority 2: Model Capacity Test (High Impact)
- Test larger model (350M+ params) with proven conservative parameters
- Baseline evaluation of untrained larger model capabilities

### Priority 3: Dataset Analysis (Medium Impact)
- Compare training vs eval set composition
- Identify potential data quality or balance issues
- Create Move 1-focused training datasets

**Bottom Line**: Hyperparameter sweeps have reached diminishing returns. Focus must shift to model architecture, data quality, and curriculum learning to break through the 70% eval accuracy ceiling.
